pip/pip-278.md

```diff
@@ -0,0 +1,156 @@
+
+For example, current topic compaction can work with pulsar format data in KoP, but it can't work with Kafka format data since the data written to the entry is in Kafka format. 
+The Pulsar compactor doesn't aware of Kafka format data. And it doesn't make sense to support Kafka format data handling in Pulsar. We need implement a pluggable compactor to support Kafka format data handling in KoP.
+
+Another long-term consideration is that we may need to support write the compacted data anywhere, S3, in columnar format, or even partitioning.
+
+So we need to make  the whole topic compaction service (include Write API & Read API) pluggable to support more customize compaction service implementation.
+
+# Goals
+
+## In Scope
+
+* Abstract topic compaction service interface and support topic compaction service pluggable.
+
+* Migrate the current implementation to a new interface implementation.
+
+* Makes existing tests compatible with new implementations.
+
+## Out of Scope
+
+* For CompactorMetrics, keep the current implementation and don't define related method in topic compaction service interface, in the future will use the `Otel` interface or other mertics API instead.
+
+* For `StrategicTwoPhaseCompactor`, it's out of the scope for regular compaction. It's only used for the load balancer. So it won't change.
+
+
+# High Level Design
+
+To make the whole topic compaction service pluggable, we need to abstract `TopicCompactionService` interface, it can provide the capability that the compactor has and provide the read API to read entries from compacted data.
+
+We should combine `CompactedTopicImpl` and `TwoPhaseCompactor` to pulsar implementation of topic compaction service and make behavior with the current implementation consistent.
+
+Class Diagram of core class:
+![pip-278-1.png](images/pip-278-1.png)
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* Define a standard TopicCompactionService interface.
+
+  ```java
+  public interface TopicCompactionService {
+      /**
+       * Compact the topic.
+       * Topic Compaction is a key-based retention mechanism. it will keep the most recent value for a given key and
+       * user will read compacted data from TopicCompactionService.
+       *
+       * @return a future that will be completed when the compaction is done.
+       */
+      CompletableFuture<Void> compact();
+
+      /**
+       * Read the compacted entries from the TopicCompactionService.
+       *
+       * @param startPosition         the position to start reading from.
+       * @param numberOfEntriesToRead the number of entries to read.
+       * @return a future that will be completed with the list of entries, this list can is null.
+       */
+      CompletableFuture<List<Entry>> readCompactedEntries(@NonNUll Position startPosition, int numberOfEntriesToRead);
```

Comment:
```suggestion
      CompletableFuture<List<Entry>> readCompactedEntries(Position startPosition, int numberOfEntriesToRead);
```

We should not add a `NonNull` annotation on a interface. If the annotation comes from lombok, it means a null check will be performed, see https://projectlombok.org/features/NonNull.

Instead, we should document the exact behavior when `startPosition` is `null` via a `@throws` or `@implNote` annotation.

Reply:
> We should not add a NonNull annotation on a interface. If the annotation comes from lombok, it means a null check will be performed, see https://projectlombok.org/features/NonNull.

`Nonnull` means [`javax.annotation.Nonnull`](https://docs.oracle.com/javaee/7/api/javax/validation/constraints/NotNull.html), it only documents this element must not be null